Enforce a consistent format for Vocabulary keys

Hamms opened the pull request:
Also update existing keys to conform to new validation.

## Testing story

Added some unit tests

## PR Checklist:

<!--
  The final step! Before you create your PR, double-check that everything is in order.
  Change [ ] to [X] during creation to check boxes.
-->

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

davidsbailey commented:
@tess323 @bethanyaconnor @dmcavoy any opinions on changing vocab key syntax from dashes to underscore? do we think curriculum writers will care which one we use?

tess323 commented:
Not that I can think of

Hamms commented:
PTAL; it turns out that we actually have [a test which expects keys to not be automatically uniquified](https://github.com/code-dot-org/code-dot-org/blob/80136dd22fe86d04dc3cd3996f830e426cffe649/dashboard/test/controllers/vocabularies_controller_test.rb#L48-L59). Which means the extra logic I added to automatically uniquify keys is actually unwanted. Unfortunately, that logic is also required to get the script seeding test to pass; it appears this is just a side effect of the significant complexity of that test and not a property of our actual script seeding process, so I updated the test to apply the logic manually. This does mean that an already-overcomplicated test has become even more complicated, but I don't think that right now is the time to invest a lot of effort in cleaning up that test; perhaps post-April.

davidsbailey commented on dashboard/test/lib/services/script_seed_test.rb:
```diff
@@ -314,9 +314,11 @@ class ScriptSeedTest < ActiveSupport::TestCase
         lesson.vocabularies.create(
           word: 'new word',
-          key: "#{lesson.name}-vocab-3",
+          key: key,
```
it seems like using a factory to create the vocab could avoid the need for the uniquify_key method. Not suggesting you do that now, but wondering if you agree with this assessment @Hamms so that we have some ideas for how to clean this up when we have time after April deadline.

Hamms replied:
Theoretically, but I found that when I tried to specify a word here and let the key be generated a whole bunch of stuff in this test case broke in ways that were not obvious to me. Hence my complaints about this test case :)

bethanyaconnor replied:
Ah, this was failing because numbers are now disallowed right? I'm not sure what the best solution is for that, other than changing the numbers in the lesson names to letters or something

davidsbailey replied:
If it solves a problem, allowing numbers in vocab keys seems ok to me.

Hamms replied:
note that this test is not actually the one for which uniqueness is necessary; I just threw that in for consistency. It's the diff below - the one in `create_script_tree` - that's the real problem

bethanyaconnor replied:
Ah good point, though I think the issue is the same (using numbers). I agree this is out of the scope of this PR. As long as `uniquify_key` is deterministic across runs (which it is), I think this is a fine solution for now.

davidsbailey commented:
> PTAL; it turns out that we actually have [a test which expects keys to not be automatically uniquified](https://github.com/code-dot-org/code-dot-org/blob/80136dd22fe86d04dc3cd3996f830e426cffe649/dashboard/test/controllers/vocabularies_controller_test.rb#L48-L59). Which means the extra logic I added to automatically uniquify keys is actually unwanted. Unfortunately, that logic is also required to get the script seeding test to pass; it appears this is just a side effect of the significant complexity of that test and not a property of our actual script seeding process, so I updated the test to apply the logic manually. This does mean that an already-overcomplicated test has become even more complicated, but I don't think that right now is the time to invest a lot of effort in cleaning up that test; perhaps post-April.

@bethanyaconnor can you provide any context on why there is a test ensuring that we do not auto-uniquify the keys for vocab words? I am speculating here, but I could see the advantage of failing when a duplicate word is created (as the test ensures) so that we don't accidentally end up with duplicate definitions of the same word.

bethanyaconnor commented:
>  I am speculating here, but I could see the advantage of failing when a duplicate word is created (as the test ensures) so that we don't accidentally end up with duplicate definitions of the same word.

Yep -- that's the exact reason! We can change that logic if it's causing too many problems

Hamms commented:
I wouldn't say it's causing problems; in fact, I much prefer a solution that _doesn't_ try to do anything super fancy with automatic uniqueness.

I'd say what's causing problems is the script seeding test, which right now raises all kinds of errors if there isn't some way to handle that automatically. Currently in this PR, the fancy automatic uniqueness logic is scoped to just that one test, and I for one am fine with that.

davidsbailey commented:
> > I am speculating here, but I could see the advantage of failing when a duplicate word is created (as the test ensures) so that we don't accidentally end up with duplicate definitions of the same word.
> 
> Yep -- that's the exact reason! We can change that logic if it's causing too many problems

No, I think this is a good reason to keep this restriction. The test seems clear in the behavior it was expecting, but perhaps a comment on it would have helped make it clear why this behavior should not be changed.

davidsbailey commented on dashboard/app/models/vocabulary.rb:
```diff
@@ -93,6 +94,27 @@ def self.sanitize_key(key)
+    end
+
+    new_key
+  end
```
I'm still not clear on why the factory method (updated to use letters instead of numbers as the sequence) does not solve this. perhaps the script_json test fixture contains entries which conflict with the keys generated by the factory? anyway, it is fine with me to keep this method and investigate later.

Hamms replied:
the problem is that the script seed test isn't using the factory method, it's using its own logic to build keys:

https://github.com/code-dot-org/code-dot-org/blob/39e267a8b9faec70cdbdb8be37d14ab333ceeea1/dashboard/test/lib/services/script_seed_test.rb#L782-L785
 
Unfortunately, it's also using a naming scheme for the levels that results in conflicts when those names are turned into keys. I looked into making improvements to the test, but everything I tried just resulted in even more failures, each more opaque than the last.

Hamms replied:
To clarify: the factory method will indeed solve this problem in general, which is why we only need to use this method in the script_seed test, not in the `generate_key` method itself. It's just the script seeding test which needs special accommodation.

davidsbailey replied:
> I looked into making improvements to the test, but everything I tried just resulted in even more failures, each more opaque than the last.

Sorry to hear the test is being such a pain. If you want to open a work item under Q2 Tech Debt and assign it to me, I can look at it after April launch.

Hamms replied:
already done :) https://codedotorg.atlassian.net/browse/PLAT-836

davidsbailey commented on dashboard/test/models/vocabulary_test.rb:
```diff
@@ -25,4 +25,55 @@ class VocabularyTest < ActiveSupport::TestCase
+  test 'vocabulary automatically sanitizes word when turning it into key' do
+    vocab = create :vocabulary,
+      word: 'Some !! ßtring 123 ,with, "ILLEGAL" _characters_.',
+      key: nil
```
why is it necessary to specify key here, given that it gets auto generated from the word?

Hamms replied:
the factory attempts to generate its own key rather than generating it from the word; passing in `nil` here prevents the factory from doing so which allows us to actually use the "generate from word" bit
